File: modules/benchmarks/arrow-scala-benchmarks/src/main/scala/cats/Delay.scala
package arrow.benchmarks.effects.scala.cats

import cats.effect.IO

object Delay {

  def ioDelayLoop(size: Int, i: Int): IO[Int] =
    IO(i).flatMap { j =>
      if (j > size) IO(j) else ioDelayLoop(size, j + 1)
    }

  def unsafeIODelayLoop(size: Int, i: Int): Int =
    ioDelayLoop(size, i).unsafeRunSync()

}

File: modules/benchmarks/arrow-kio-benchmarks/src/main/kotlin/kio/Delay.kt
package arrow.benchmarks.effects.kio

import it.msec.kio.UIO
import it.msec.kio.effect
import it.msec.kio.flatMap
import it.msec.kio.runtime.Runtime

object Delay {

  private fun kioDelayLoop(size: Int, i: Int): UIO<Int> =
    effect { i }.flatMap { j ->
      if (j > size) effect { j } else kioDelayLoop(size, j + 1)
    }

  fun unsafeIODelayLoop(size: Int, i: Int): Int =
    Runtime.unsafeRunSyncAndGet(kioDelayLoop(size, i))
}
